bluefog/run/interactive_run.py

```diff
@@ -0,0 +1,321 @@
+        mpi_ipengine_command = (
+            'mpirun --allow-run-as-root '
+            '-np {num_proc} {hosts_arg} '
+            '-bind-to none -map-by slot '
+            '-mca pml ob1 '
+            '{ssh_port_arg} {tcp_intf_arg} '
+            '{extra_flags} {nccl_socket_intf_arg} '
+            '{command}'
+            .format(num_proc=args.np,
+                    hosts_arg=hosts_arg,
+                    ssh_port_arg=ssh_port_arg,
+                    tcp_intf_arg=tcp_intf_arg,
+                    nccl_socket_intf_arg=nccl_socket_intf_arg,
+                    extra_flags=extra_flags,
+                    env=' '.join('-x %s' % key for key in env.keys()
+                                    if env_util.is_exportable(key)),
+                    command=ipengine_command)
+        )
+        subprocess.run(ipengine_command, shell=True,
+                        env=env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    else:
+        subprocess.run(ipcluster_stop_command, shell=True,
+                       env=env, capture_output=True)
+        _maybe_kill_ipcontroller_process(args.profile)
+
+
+def main():
+    args = parse_args()
+
+    if args.version:
+        print(bluefog.__version__)
+        exit(0)
+
+    hosts_arg, all_host_names = network_util.get_hosts_arg_and_hostnames(args)
+    remote_host_names = network_util.filter_local_addresses(all_host_names)
+
+    if not env_util.is_open_mpi_installed():
+        raise Exception(
+            'ibfrun convenience script currently only supports Open MPI.\n'
+            'Please Install Open MPI 4.0.0+ and re-install Bluefog.')
+
+    if not env_util.is_ipyparallel_installed():
+        raise Exception(
+            'ibfrun is based on the ipyparallel package. Please install it in your\n'
+            'system like `pip install ipyparallel` first, then run ibfrun again.'
+        )
+
+    env = os.environ.copy()
+    env['BLUEFOG_CYCLE_TIME'] = str(20)  # Increase the cycle time
+    if len(args.command) != 1 or args.command[0] not in ("start", "stop"):
+        raise ValueError("The last command has to be either 'start' or 'stop', but it is "
+                         "{} now.".format(args.command))
+    command = args.command[0]
+
+    # TODO(ybc) How to stop it properly?
+    # In multiple machine env, we alose need to remove engine.json and client.json file.
+    ipcluster_stop_command = "ipcluster stop --profile {profile}".format(
+        profile=args.profile)
+
+    try:
+        if not remote_host_names:
+            local_machine_launch(args, env, command, ipcluster_stop_command)
+        else:
+            multiple_machines_launch(args, env, all_host_names=all_host_names,
+                                     hosts_arg=hosts_arg,
+                                     remote_host_names=remote_host_names,
+                                     command=command,
+                                     ipcluster_stop_command=ipcluster_stop_command)
```

Comment:
If possible, move `ipcluster_stop_command` out from `multiple_machines_launch` and `local_machine_launch`, and let them only handle `ibfrun start` logic.

Reply:
With the sub-command change, I split the logic of start and stop completely. Now the code is much clear.